python/lsst/daf/butler/core/utils.py

```diff
@@ -147,3 +147,31 @@ def doImport(pythonType):
     importedClass = doImport(importClassString)
     pythonType = getattr(importedClass, pythonTypeTokenList[-1])
     return pythonType
+
+
+def getInstanceOf(typeOrName):
+    """Given the type name or a type, instantiate an object of that type.
+
+    If a type name is given, an attempt will be made to import the type.
+
+    Parameters
+    ----------
+    typeOrName : `str` or Python class
+        A string describing the Python class to load or a Python type.
+    """
+    if isinstance(typeOrName, str):
+        cls = doImport(typeOrName)
+    else:
+        cls = typeOrName
+    return cls()
+
+
+class Singleton(type):
+    """Metaclass to convert a class to a Singleton."""
+
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):  # noqa N805
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
```

Comment:
you only ever get the first thing you constructed back. The `**kwargs` are there to allow this metaclass to be generally useful. I never call the `StorageClassFactory` constructor with arguments now because it no longer makes any sense because you don't know if you are the first one. I could remove the arguments from this definition if you want to force our uses of `Singleton` to only work if there are no arguments.

Reply:
Yes, I think that would be safer.